# ACA2 Tianyi Dai

In [81]:
import pandas as pd
import numpy as np
from sklearn import preprocessing

In [82]:
aca2=pd.read_csv("aca2_dataset_training.csv")
aca2_valid=pd.read_csv("aca2_dataset_validation.csv")

In [83]:
aca2

UNIQUEID SCHOOL Class  GRADE CODER  STUDENTID  Gender  OBSNUM  \
0         34880      A   T8T      2     Z     965160       1      12   
1         32344      B   T9U      3     Z      11665       1      11   
2         14774      B   T9Q      0     Z     187441       0      21   
3         19468      A   T8S      1     Z     402837       0      17   
4         31570      B   T9T      2     Z     300647       1       2   
...         ...    ...   ...    ...   ...        ...     ...     ...   
22179     30762      B   T9Q      0     Z     749212       1       5   
22180     14611      B   T9Q      0     Z      87701       0      12   
22181     20432      A  T8VQ      4     Z     130859       0       6   
22182     38264      E   T6Q      0     Y     874565       1      10   
22183     21734      A   T8R      4     Z     468563       0       3   

       totalobs-forsession     Activity ONTASK  TRANSITIONS  NumACTIVITIES  \
0                      224  Wholecarpet      Y            1              2   
1                      225   Individual      Y            2              3   
2                      374   Smallgroup      Y            3              4   
3                      320   Wholedesks      Y            5              6   
4                       35   Wholedesks      Y            2              3   
...                    ...          ...    ...          ...            ...   
22179                   95  Wholecarpet      Y            2              3   
22180                  211   Smallgroup      N            3              4   
22181                   95   Smallgroup      N            1              2   
22182                  147   Wholedesks      Y            1              2   
22183                   49   Wholedesks      Y            0              1   

       FORMATchanges  NumFORMATS     Obsv/act  Transitions/Durations  \
0                  1           2  1390.500000               0.000664   
1                  2           3   890.333333               0.002545   
2                  1           2   770.500000               0.001439   
3                  0           1   592.333333               0.001407   
4                  1           2   870.000000               0.005882   
...              ...         ...          ...                    ...   
22179              2           3   969.000000               0.002959   
22180              1           2   770.500000               0.001439   
22181              0           1  1060.000000               0.000472   
22182              1           2  1204.500000               0.000838   
22183              0           1  2076.000000               0.000000   

       Total Time  
0            1027  
1             213  
2            2078  
3            3229  
4             260  
...           ...  
22179         652  
22180         708  
22181         362  
22182          60  
22183         803  

[22184 rows x 18 columns]

In [84]:
aca2_valid

UNIQUEID SCHOOL Class  GRADE CODER  STUDENTID  Gender  OBSNUM  \
0        20774      A  T8VQ      4     Z      15210       0       1   
1        30961      B   T9S      1     Z     165165       0       1   
2        35767      A   T8R      4     Z     188548       1       1   
3        19990      A   T8T      2     Z     224646       0       1   
4        18668      A   T8S      1     Z     402837       0       1   
...        ...    ...   ...    ...   ...        ...     ...     ...   
5542     21660      A   T8R      4     Z     857436       0      31   
5543     21662      A   T8R      4     Z     402069       1      31   
5544     21665      A   T8R      4     Z      94671       1      31   
5545     21682      A   T8R      4     Z     762145       1      32   
5546     21683      A   T8R      4     Z     753929       1      32   

      totalobs-forsession     Activity ONTASK  TRANSITIONS  NumACTIVITIES  \
0                       1  Wholecarpet      Y            1              2   
1                       1   Wholedesks      Y            3              4   
2                       1  Wholecarpet      Y            3              4   
3                       1  Wholecarpet      N            2              3   
4                       1   Individual      Y            3              4   
...                   ...          ...    ...          ...            ...   
5542                  487      Testing      Y            0              1   
5543                  489      Testing      Y            0              1   
5544                  492      Testing      N            0              1   
5545                  509      Testing      Y            0              1   
5546                  510      Testing      Y            0              1   

      FORMATchanges  NumFORMATS  Obsv/act  Transitions/Durations  Total Time  
0                 1           2   1677.00               0.000446           0  
1                 3           4    826.00               0.003672           0  
2                 2           3    862.75               0.004658           0  
3                 0           1   1073.00               0.000621           0  
4                 2           3   1106.75               0.002841           0  
...             ...         ...       ...                    ...         ...  
5542              0           1   2530.00               0.000000        2367  
5543              0           1   2530.00               0.000000        2375  
5544              0           1   2530.00               0.000000        2389  
5545              0           1   2530.00               0.000000        2522  
5546              0           1   2530.00               0.000000        2525  

[5547 rows x 18 columns]

In [85]:
aca2['ONTASK'].value_counts()

Y    14938
N     7246
Name: ONTASK, dtype: int64

In [86]:
aca2_valid['ONTASK'].value_counts()

Y    3698
N    1849
Name: ONTASK, dtype: int64

the proportion of on-task behaviour was fairly high.  
For the training set, Y=14938 and N=7246; for the validation set, Y=3698 and N=1849. The ratios are all Y:N=2:1.

In [87]:
aca2.describe()

UNIQUEID         GRADE      STUDENTID        Gender        OBSNUM  \
count  22184.000000  22184.000000   22184.000000  22184.000000  22184.000000   
mean   28256.558556      2.056076  506965.948521      0.506446      9.621168   
std     8030.006461      1.503220  283257.429497      0.499970      5.702232   
min    14400.000000      0.000000    1123.000000      0.000000      1.000000   
25%    21276.750000      1.000000  264220.000000      0.000000      5.000000   
50%    28263.500000      2.000000  514301.000000      1.000000      9.000000   
75%    35231.250000      4.000000  743450.000000      1.000000     14.000000   
max    42130.000000      4.000000  999979.000000      1.000000     32.000000   

       totalobs-forsession   TRANSITIONS  NumACTIVITIES  FORMATchanges  \
count         22184.000000  22184.000000   22184.000000   22184.000000   
mean            170.693608      2.383114       3.383114       1.534304   
std             105.788938      1.315163       1.315163       1.229399   
min               1.000000      0.000000       1.000000       0.000000   
25%              82.000000      1.000000       2.000000       1.000000   
50%             165.000000      2.000000       3.000000       1.000000   
75%             248.000000      3.000000       4.000000       2.000000   
max             511.000000      6.000000       7.000000       5.000000   

         NumFORMATS      Obsv/act  Transitions/Durations    Total Time  
count  22184.000000  22184.000000           22184.000000  22184.000000  
mean       2.534304    973.525214               0.003159    774.587811  
std        1.229399    454.538354               0.010309    665.169774  
min        1.000000    387.000000               0.000000      0.000000  
25%        2.000000    721.200000               0.000839    252.000000  
50%        2.000000    876.250000               0.001513    586.500000  
75%        3.000000   1106.750000               0.003268   1121.000000  
max        6.000000   2735.000000               0.666667   3554.000000

## Pre-processing 

###  Label Transformation

In [88]:
label=preprocessing.LabelEncoder()
label.fit(np.unique(list(aca2['Activity'].values)))
aca2['Activity']=label.transform(list(aca2['Activity'].values))
label.fit(np.unique(list(aca2['SCHOOL'].values)))
aca2['SCHOOL']=label.transform(list(aca2['SCHOOL'].values))
label.fit(np.unique(list(aca2['Class'].values)))
aca2['Class']=label.transform(list(aca2['Class'].values))

In [89]:
aca2

UNIQUEID  SCHOOL  Class  GRADE CODER  STUDENTID  Gender  OBSNUM  \
0         34880       0     15      2     Z     965160       1      12   
1         32344       1     20      3     Z      11665       1      11   
2         14774       1     17      0     Z     187441       0      21   
3         19468       0     14      1     Z     402837       0      17   
4         31570       1     19      2     Z     300647       1       2   
...         ...     ...    ...    ...   ...        ...     ...     ...   
22179     30762       1     17      0     Z     749212       1       5   
22180     14611       1     17      0     Z      87701       0      12   
22181     20432       0     16      4     Z     130859       0       6   
22182     38264       4      5      0     Y     874565       1      10   
22183     21734       0     13      4     Z     468563       0       3   

       totalobs-forsession  Activity ONTASK  TRANSITIONS  NumACTIVITIES  \
0                      224         4      Y            1              2   
1                      225         1      Y            2              3   
2                      374         2      Y            3              4   
3                      320         5      Y            5              6   
4                       35         5      Y            2              3   
...                    ...       ...    ...          ...            ...   
22179                   95         4      Y            2              3   
22180                  211         2      N            3              4   
22181                   95         2      N            1              2   
22182                  147         5      Y            1              2   
22183                   49         5      Y            0              1   

       FORMATchanges  NumFORMATS     Obsv/act  Transitions/Durations  \
0                  1           2  1390.500000               0.000664   
1                  2           3   890.333333               0.002545   
2                  1           2   770.500000               0.001439   
3                  0           1   592.333333               0.001407   
4                  1           2   870.000000               0.005882   
...              ...         ...          ...                    ...   
22179              2           3   969.000000               0.002959   
22180              1           2   770.500000               0.001439   
22181              0           1  1060.000000               0.000472   
22182              1           2  1204.500000               0.000838   
22183              0           1  2076.000000               0.000000   

       Total Time  
0            1027  
1             213  
2            2078  
3            3229  
4             260  
...           ...  
22179         652  
22180         708  
22181         362  
22182          60  
22183         803  

[22184 rows x 18 columns]

In [90]:
label.fit(np.unique(list(aca2_valid['Activity'].values)))
aca2_valid['Activity']=label.transform(list(aca2_valid['Activity'].values))
label.fit(np.unique(list(aca2_valid['SCHOOL'].values)))
aca2_valid['SCHOOL']=label.transform(list(aca2_valid['SCHOOL'].values))
label.fit(np.unique(list(aca2_valid['Class'].values)))
aca2_valid['Class']=label.transform(list(aca2_valid['Class'].values))

In [91]:
aca2_valid

UNIQUEID  SCHOOL  Class  GRADE CODER  STUDENTID  Gender  OBSNUM  \
0        20774       0     16      4     Z      15210       0       1   
1        30961       1     18      1     Z     165165       0       1   
2        35767       0     13      4     Z     188548       1       1   
3        19990       0     15      2     Z     224646       0       1   
4        18668       0     14      1     Z     402837       0       1   
...        ...     ...    ...    ...   ...        ...     ...     ...   
5542     21660       0     13      4     Z     857436       0      31   
5543     21662       0     13      4     Z     402069       1      31   
5544     21665       0     13      4     Z      94671       1      31   
5545     21682       0     13      4     Z     762145       1      32   
5546     21683       0     13      4     Z     753929       1      32   

      totalobs-forsession  Activity ONTASK  TRANSITIONS  NumACTIVITIES  \
0                       1         4      Y            1              2   
1                       1         5      Y            3              4   
2                       1         4      Y            3              4   
3                       1         4      N            2              3   
4                       1         1      Y            3              4   
...                   ...       ...    ...          ...            ...   
5542                  487         3      Y            0              1   
5543                  489         3      Y            0              1   
5544                  492         3      N            0              1   
5545                  509         3      Y            0              1   
5546                  510         3      Y            0              1   

      FORMATchanges  NumFORMATS  Obsv/act  Transitions/Durations  Total Time  
0                 1           2   1677.00               0.000446           0  
1                 3           4    826.00               0.003672           0  
2                 2           3    862.75               0.004658           0  
3                 0           1   1073.00               0.000621           0  
4                 2           3   1106.75               0.002841           0  
...             ...         ...       ...                    ...         ...  
5542              0           1   2530.00               0.000000        2367  
5543              0           1   2530.00               0.000000        2375  
5544              0           1   2530.00               0.000000        2389  
5545              0           1   2530.00               0.000000        2522  
5546              0           1   2530.00               0.000000        2525  

[5547 rows x 18 columns]

### Drop unrelated variables 

In [92]:
aca2=aca2.drop('UNIQUEID', axis = 1)
aca2=aca2.drop('CODER', axis = 1)
aca2=aca2.drop('TRANSITIONS', axis = 1)
aca2=aca2.drop('FORMATchanges', axis = 1)
aca2=aca2.drop('STUDENTID', axis = 1)
aca2

SCHOOL  Class  GRADE  Gender  OBSNUM  totalobs-forsession  Activity  \
0           0     15      2       1      12                  224         4   
1           1     20      3       1      11                  225         1   
2           1     17      0       0      21                  374         2   
3           0     14      1       0      17                  320         5   
4           1     19      2       1       2                   35         5   
...       ...    ...    ...     ...     ...                  ...       ...   
22179       1     17      0       1       5                   95         4   
22180       1     17      0       0      12                  211         2   
22181       0     16      4       0       6                   95         2   
22182       4      5      0       1      10                  147         5   
22183       0     13      4       0       3                   49         5   

      ONTASK  NumACTIVITIES  NumFORMATS     Obsv/act  Transitions/Durations  \
0          Y              2           2  1390.500000               0.000664   
1          Y              3           3   890.333333               0.002545   
2          Y              4           2   770.500000               0.001439   
3          Y              6           1   592.333333               0.001407   
4          Y              3           2   870.000000               0.005882   
...      ...            ...         ...          ...                    ...   
22179      Y              3           3   969.000000               0.002959   
22180      N              4           2   770.500000               0.001439   
22181      N              2           1  1060.000000               0.000472   
22182      Y              2           2  1204.500000               0.000838   
22183      Y              1           1  2076.000000               0.000000   

       Total Time  
0            1027  
1             213  
2            2078  
3            3229  
4             260  
...           ...  
22179         652  
22180         708  
22181         362  
22182          60  
22183         803  

[22184 rows x 13 columns]

In [93]:
aca2_valid=aca2_valid.drop('UNIQUEID', axis = 1)
aca2_valid=aca2_valid.drop('CODER', axis = 1)
aca2_valid=aca2_valid.drop('TRANSITIONS', axis = 1)
aca2_valid=aca2_valid.drop('FORMATchanges', axis = 1)
aca2_valid=aca2_valid.drop('STUDENTID', axis = 1)
aca2_valid

SCHOOL  Class  GRADE  Gender  OBSNUM  totalobs-forsession  Activity  \
0          0     16      4       0       1                    1         4   
1          1     18      1       0       1                    1         5   
2          0     13      4       1       1                    1         4   
3          0     15      2       0       1                    1         4   
4          0     14      1       0       1                    1         1   
...      ...    ...    ...     ...     ...                  ...       ...   
5542       0     13      4       0      31                  487         3   
5543       0     13      4       1      31                  489         3   
5544       0     13      4       1      31                  492         3   
5545       0     13      4       1      32                  509         3   
5546       0     13      4       1      32                  510         3   

     ONTASK  NumACTIVITIES  NumFORMATS  Obsv/act  Transitions/Durations  \
0         Y              2           2   1677.00               0.000446   
1         Y              4           4    826.00               0.003672   
2         Y              4           3    862.75               0.004658   
3         N              3           1   1073.00               0.000621   
4         Y              4           3   1106.75               0.002841   
...     ...            ...         ...       ...                    ...   
5542      Y              1           1   2530.00               0.000000   
5543      Y              1           1   2530.00               0.000000   
5544      N              1           1   2530.00               0.000000   
5545      Y              1           1   2530.00               0.000000   
5546      Y              1           1   2530.00               0.000000   

      Total Time  
0              0  
1              0  
2              0  
3              0  
4              0  
...          ...  
5542        2367  
5543        2375  
5544        2389  
5545        2522  
5546        2525  

[5547 rows x 13 columns]

## Decision Tree

In [94]:
from sklearn.tree import DecisionTreeClassifier
aca2_tree = aca2.drop('ONTASK', axis = 1)
ONTASK = aca2['ONTASK']

In [95]:
aca2TreeModel = DecisionTreeClassifier()
aca2TreeModel.fit(aca2_tree, ONTASK)

DecisionTreeClassifier()

In [96]:
ontask_pred = aca2TreeModel.predict(aca2_tree)
performance = [item in ontask_pred for item in ONTASK]
print('The accuracy is', sum(performance)/len(performance)*100, '%')

The accuracy is 100.0 %


fit the model on validation data and evaluation the accuarcy

In [114]:
aca2_valid_tree = aca2_valid.drop('ONTASK', axis = 1)
ONTASK_valid = aca2_valid['ONTASK']

In [98]:
ontask_valid_pred = aca2TreeModel.predict(aca2_valid_tree)

In [99]:
from sklearn.metrics import accuracy_score
ac_tree = accuracy_score(ONTASK_valid, ontask_valid_pred)
print("The accuracy for decision tree", ac_tree*100, "%")

The accuracy for decision tree 60.176672074995494 %


## Naive Bayes

In [107]:
from sklearn.naive_bayes import GaussianNB
aca2_NB = aca2.drop('ONTASK', axis = 1)
ONTASK = aca2['ONTASK']

In [108]:
aca2NBModel = GaussianNB()
aca2NBModel.fit(aca2_NB, ONTASK)

GaussianNB()

In [109]:
ontask_pred = aca2NBModel.predict(aca2_NB)
performance = [item in ontask_pred for item in ONTASK]
print('The accuracy is', sum(performance)/len(performance)*100, '%')

The accuracy is 100.0 %


fit the model on validation data and evaluation the accuarcy

In [110]:
aca2_valid_NB = aca2_valid.drop('ONTASK', axis = 1)
ONTASK_valid = aca2_valid['ONTASK']

In [111]:
ontask_valid_pred = aca2NBModel.predict(aca2_valid_tree)

In [113]:
from sklearn.metrics import accuracy_score
ac_NB = accuracy_score(ONTASK_valid, ontask_valid_pred)
print("The accuracy for naive bayes", ac_NB*100, "%")

The accuracy for naive bayes 66.66666666666666 %


## Analysis 

The accuarcy of Naive Bayes is much higher than Decision Tree. Using different vairables will lead to different result, so some unrelated variables should be droped, or they will make the model overfitting.  
  
For this dataset, the on/off-task behavior may related to "School", "class", "grade", "gender".  
Class: on-task behaviour was weakly related to learning outcomes.  
Grade: The higher the grade is, the less off-task behavior in class.  
Gender: Female students tend to exhibit more on-task behaviour than male students in elementary school.  
  
Therefore, these findings highlight the complex relationship between time on-task and learning, and call for more research to identify factors or conditions where more time on-task yields better learning outcomes.